### PR TITLE
Bad SYSTEM for DTD DocType and Entity breaks the XML validation

### LINKED ISSUE #1169: Bad SYSTEM for DTD DocType and Entity breaks the XML validation
See https://github.com/redhat-developer/vscode-xml/issues/647

### PR DESCRIPTION
**Pull request by angelozerr**
Bad SYSTEM for DTD DocType and Entity breaks the XML validation

Fixes #1169

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		}
+	}
+
+	@Override
+	public String setupCurrentEntity(String name, XMLInputSource xmlInputSource, boolean literal, boolean isExternal)
+			throws IOException, XNIException {
+		try {
+			return super.setupCurrentEntity(name, xmlInputSource, literal, isExternal);
+		} catch (Exception e) {
+			// DTD DOCTYPE/ENTITY SYSTEM doesn't exists:
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java:77]: doesn't exist

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+			// 1. report the error
+			reportError(xmlInputSource.getSystemId(), e);
+			// 2. report an empty String to avoid breaking syntax validation
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java:80]: return an empty string

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		hasProblemsWithReferencedDTD = true;
+		errorReporter.reportError(NULL_LOCATOR, XMLModelMessageFormatter.XML_MODEL_DOMAIN,
+				DTDErrorCode.dtd_not_found.getCode(), new Object[] { null, location }, XMLErrorReporter.SEVERITY_ERROR,
+				e);
+	}
+
+	public void dispose() {
+		if (hasProblemsWithReferencedDTD) {
+			// one of DTD which where cached has an error
+			// remove all cached DTD during the validation from the grammar ppol.
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java:97]: all DTDs cached during

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+	@Override
+	public XMLInputSource resolveEntity(XMLResourceIdentifier resourceIdentifier) throws IOException, XNIException {
+		if (resourceIdentifier instanceof XSDDescription) {
+			return super.resolveEntity(resourceIdentifier);
+		}
+		try {
+			return super.resolveEntity(resourceIdentifier);
+		} catch (Exception e) {
+			// DTD DOCTYPE/ENTITY SYSTEM doesn't exists:
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+			// 1. report the error
+			reportError(resourceIdentifier.getLiteralSystemId(), e);
+			// 2. report an empty String to avoid breaking syntax validation
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	public void dispose() {
+		if (hasProblemsWithReferencedDTD) {
+			// one of DTD which where cached has an error
+			// remove all cached DTD during the validation from the grammar ppol.
+			grammarPool.clear();
+		}
+	}
+
+	/**
+	 * The error code 'dtd_not_found' don't use the locator, but to avoid having
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/diagnostics/LSPXMLEntityManager.java]: fixed

**Comment by fbricon** [comment]
Thanks @angelozerr!

**Pull request merged**